crates/core/src/event.rs

```diff
@@ -13,21 +14,86 @@ use thiserror::Error;
 
 use crate::borsh::{BorshDeserialize, BorshSerialize};
 use crate::collections::HashMap;
-use crate::ethereum_structs::{BpTransferStatus, EthBridgeEvent};
+use crate::ethereum_structs::EthBridgeEvent;
 use crate::ibc::IbcEvent;
 
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __event_type_impl {
+    ($domain:ty) => {
+        <$domain as $crate::event::EventToEmit>::DOMAIN
+    };
+    ($domain:ty, $($subdomain:expr),*) => {
+        ::konst::string::str_join!(
+            "/",
+            &[
+                $crate::__event_type_impl!($domain),
+                $($subdomain),*
+            ],
+        )
+    };
+    // TODO: remove these variants of the macro
+    ($domain:expr) => {
+        $domain
+    };
+    ($domain:expr, $($subdomain:expr),*) => {
+        ::konst::string::str_join!(
+            "/",
+            &[
+                $crate::__event_type_impl!($domain),
+                $($subdomain),*
+            ],
+        )
+    };
+}
+
+/// Instantiate a new [`EventType`] in const contexts. Mostly
+/// useful to define new event types in the protocol.
+///
+/// # Example
+///
+/// ```ignore
+/// const RELAYED: EventType = event_type!(EthBridgeEvent, "bridge-pool", "relayed");
+/// ```
+#[macro_export]
+macro_rules! event_type {
+    ($($tt:tt)*) => {
+        $crate::event::EventType::new($crate::__event_type_impl!($($tt)*))
+    };
+}
+
+/// An event to be emitted in Namada.
+pub trait EventToEmit: Into<Event> {
```

Comment:
I'm probably losing my mind, but I went to find the `From<IbcEvent> for Event` and couldn't find it anywhere

Reply:
line 399 of this file

Reply:
I thought the point of all this traits was to localize implementations to the appropriate crate. Seems funny to do this here.

Reply:
it is yeah, but core still depends on `IbcEvent` :( eventually we want to move `IbcEvent` to the `ibc` crate